contracts/sg721/src/contract.rs

```diff
@@ -96,8 +99,74 @@ pub fn execute(
     env: Env,
     info: MessageInfo,
     msg: ExecuteMsg,
-) -> Result<Response, BaseError> {
-    Sg721Contract::default().execute(deps, env, info, msg)
+) -> Result<Response, ContractError> {
+    match msg {
+        ExecuteMsg::Freeze {} => execute_freeze(deps, info),
+        ExecuteMsg::UpdateTokenURIs { base_token_uri } => {
+            execute_update_token_uris(deps, env, info, base_token_uri)
+        }
+        _ => Sg721Contract::default()
+            .execute(deps, env, info, msg.into())
+            .map_err(ContractError::from),
+    }
+}
+
+fn execute_freeze(deps: DepsMut, info: MessageInfo) -> Result<Response, ContractError> {
+    let sg721_contract = Sg721Contract::default();
+    let minter = sg721_contract.minter.load(deps.storage)?;
+
+    if info.sender != minter {
+        return Err(ContractError::Unauthorized {});
+    }
+
+    FROZEN.save(deps.storage, &true)?;
+
+    Ok(Response::new().add_attribute("action", "freeze"))
+}
+fn execute_update_token_uris(
+    deps: DepsMut,
+    env: Env,
+    info: MessageInfo,
+    base_token_uri: String,
+) -> Result<Response, ContractError> {
+    let sg721_contract = Sg721Contract::default();
+    let minter = sg721_contract.minter.load(deps.storage)?;
+    let frozen = FROZEN.load(deps.storage)?;
+
+    if info.sender != minter {
+        return Err(ContractError::Unauthorized {});
+    }
+
+    if frozen {
+        return Err(ContractError::Frozen {});
+    }
+
+    // iterate through all tokens
+    let all_tokens_msg = QueryMsg::AllTokens {
+        start_after: None,
+        limit: None,
+    };
+    let tokens_response: TokensResponse =
+        from_binary(&query(deps.as_ref(), env, all_tokens_msg).unwrap()).unwrap();
```

Comment:
This query is paginated, not sure how to handle it? Also code smell.

It's better to do `sg721` level change to store `base_token_uri` and return individual token uri as `{base_token_uri}/{token_id}`. Maybe as a separate issue.

Reply:
I think one way to do this is to make the `baseUri` optional. Then if it exists, update it only and treat the collection as if it's a generative collection. If `baseUri` doesn't exist, then use the individual `tokenUri`s.